NumPy 2.3 support

Testing Numba 0.61.2 locally with NumPy 2.3 results in the following failures:

```
======================================================================
ERROR: test_global_record (numba.tests.test_globals.TestGlobals.test_global_record)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/tests/test_globals.py", line 215, in test_global_record
    res = global_record_func(x)
          ^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/dispatcher.py", line 443, in _compile_for_args
    raise e
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/dispatcher.py", line 376, in _compile_for_args
    return_val = self.compile(tuple(argtypes))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/dispatcher.py", line 904, in compile
    cres = self._compiler.compile(args, return_type)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/dispatcher.py", line 80, in compile
    status, retval = self._compile_cached(args, return_type)
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/dispatcher.py", line 94, in _compile_cached
    retval = self._compile_core(args, return_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/dispatcher.py", line 107, in _compile_core
    cres = compiler.compile_extra(self.targetdescr.typing_context,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/compiler.py", line 739, in compile_extra
    return pipeline.compile_extra(func)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/compiler.py", line 439, in compile_extra
    return self._compile_bytecode()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/compiler.py", line 505, in _compile_bytecode
    return self._compile_core()
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/compiler.py", line 481, in _compile_core
    raise e
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/compiler.py", line 473, in _compile_core
    pm.run(self.state)
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/compiler_machinery.py", line 363, in run
    raise e
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/compiler_machinery.py", line 356, in run
    self._runPass(idx, pass_inst, state)
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/compiler_lock.py", line 35, in _acquire_compile_lock
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/compiler_machinery.py", line 311, in _runPass
    mutated |= check(pss.run_pass, internal_state)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/compiler_machinery.py", line 272, in check
    mangled = func(compiler_state)
              ^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/typed_passes.py", line 468, in run_pass
    lower.lower()
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/lowering.py", line 193, in lower
    self.lower_normal_function(self.fndesc)
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/lowering.py", line 232, in lower_normal_function
    entry_block_tail = self.lower_function_body()
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/lowering.py", line 262, in lower_function_body
    self.lower_block(block)
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/lowering.py", line 276, in lower_block
    self.lower_inst(inst)
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/lowering.py", line 462, in lower_inst
    val = self.lower_assign(ty, inst)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/lowering.py", line 668, in lower_assign
    res = self.context.get_constant_generic(self.builder, ty,
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/core/base.py", line 506, in get_constant_generic
    return impl(self, builder, ty, val)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/np/arrayobj.py", line 3337, in constant_record
    val = lty(bytearray(pyval.tostring()))
                        ^^^^^^^^^^^^^^^^
AttributeError: 'numpy.ndarray' object has no attribute 'tostring'

======================================================================
ERROR: test_argpartition_boolean_inputs (numba.tests.test_np_functions.TestNPFunctions.test_argpartition_boolean_inputs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/tests/test_np_functions.py", line 2948, in test_argpartition_boolean_inputs
    self.argpartition_sanity_check(pyfunc, cfunc, d, kth)
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/tests/test_np_functions.py", line 2296, in argpartition_sanity_check
    expected = pyfunc(a, kth)
               ^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/tests/test_np_functions.py", line 308, in argpartition
    return np.argpartition(a, kth)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numpy/_core/fromnumeric.py", line 961, in argpartition
    return _wrapfunc(a, 'argpartition', kth, axis=axis, kind=kind, order=order)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numpy/_core/fromnumeric.py", line 57, in _wrapfunc
    return bound(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^^
ValueError: Booleans unacceptable as partition index

======================================================================
ERROR: test_partition_boolean_inputs (numba.tests.test_np_functions.TestNPFunctions.test_partition_boolean_inputs)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/tests/test_np_functions.py", line 2940, in test_partition_boolean_inputs
    self.partition_sanity_check(pyfunc, cfunc, d, kth)
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/tests/test_np_functions.py", line 2283, in partition_sanity_check
    expected = pyfunc(a, kth)
               ^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/tests/test_np_functions.py", line 304, in partition
    return np.partition(a, kth)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numpy/_core/fromnumeric.py", line 867, in partition
    a.partition(kth, axis=axis, kind=kind, order=order)
ValueError: Booleans unacceptable as partition index

======================================================================
FAIL: test_record_type_equiv (numba.tests.test_types.TestRecordDtype.test_record_type_equiv)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/gmarkall/miniforge3/envs/numba-np-23/lib/python3.12/site-packages/numba/tests/test_types.py", line 570, in test_record_type_equiv
    self.assertEqual(art2.dtype.dtype, rec_ty)
AssertionError: dtype([('a', '<i4'), ('b', '<f4')]) != dtype(Record(a[type=int32;offset=0],b[type=float32;offset=4];8;False))
```

Doesn't look like there's anything major, but each of these needs a look at.

--- a/.github/workflows/conda_workflow_matrix.json
+++ b/.github/workflows/conda_workflow_matrix.json
@@ -14,6 +14,7 @@
       { "python-version": "3.12", "numpy_test": "1.26" },
       { "python-version": "3.12", "numpy_test": "2.0" },
       { "python-version": "3.12", "numpy_test": "2.2" },
-      { "python-version": "3.13", "numpy_test": "2.2" }
+      { "python-version": "3.13", "numpy_test": "2.2" },
+      { "python-version": "3.13", "numpy_test": "2.3" }
     ]
 }

--- a/.github/workflows/wheel_workflow_matrix.json
+++ b/.github/workflows/wheel_workflow_matrix.json
@@ -14,6 +14,7 @@
         { "python-version": "3.12", "numpy_test": "1.26" },
         { "python-version": "3.12", "numpy_test": "2.0" },
         { "python-version": "3.12", "numpy_test": "2.2" },
-        { "python-version": "3.13", "numpy_test": "2.2" }
+        { "python-version": "3.13", "numpy_test": "2.2" },
+        { "python-version": "3.13", "numpy_test": "2.3" }
       ]
 }

--- a/docs/upcoming_changes/10147.highlight.rst
+++ b/docs/upcoming_changes/10147.highlight.rst
@@ -1,0 +1,4 @@
+Support for NumPy 2.3
+---------------------
+
+Numba now supports NumPy 2.3.

--- a/numba/np/arraymath.py
+++ b/numba/np/arraymath.py
@@ -1885,7 +1885,10 @@ def np_partition(a, kth):
         raise NumbaTypeError(msg)
 
     kthdt = getattr(kth, 'dtype', kth)
-    kth_types = types.Integer if numpy_version >= (2, 3) else (types.Boolean, types.Integer)
+    if numpy_version >= (2, 3):
+      kth_types = types.Integer 
+    else:
+      kth_types = (types.Boolean, types.Integer)
     if not isinstance(kthdt, kth_types):
         # bool gets cast to int subsequently
         raise NumbaTypeError('Partition index must be integer')
@@ -1912,7 +1915,10 @@ def np_argpartition(a, kth):
         raise NumbaTypeError(msg)
 
     kthdt = getattr(kth, 'dtype', kth)
-    kth_types = types.Integer if numpy_version >= (2, 3) else (types.Boolean, types.Integer)
+    if numpy_version >= (2, 3):
+      kth_types = types.Integer 
+    else:
+      kth_types = (types.Boolean, types.Integer)
     if not isinstance(kthdt, kth_types):
         # bool gets cast to int subsequently
         raise NumbaTypeError('Partition index must be integer')

--- a/numba/np/arraymath.py
+++ b/numba/np/arraymath.py
@@ -1885,7 +1885,8 @@ def np_partition(a, kth):
         raise NumbaTypeError(msg)
 
     kthdt = getattr(kth, 'dtype', kth)
-    if not isinstance(kthdt, (types.Boolean, types.Integer)):
+    kth_types = types.Integer if numpy_version >= (2, 3) else (types.Boolean, types.Integer)
+    if not isinstance(kthdt, kth_types):
         # bool gets cast to int subsequently
         raise NumbaTypeError('Partition index must be integer')
 
@@ -1911,7 +1912,8 @@ def np_argpartition(a, kth):
         raise NumbaTypeError(msg)
 
     kthdt = getattr(kth, 'dtype', kth)
-    if not isinstance(kthdt, (types.Boolean, types.Integer)):
+    kth_types = types.Integer if numpy_version >= (2, 3) else (types.Boolean, types.Integer)
+    if not isinstance(kthdt, kth_types):
         # bool gets cast to int subsequently
         raise NumbaTypeError('Partition index must be integer')
 

--- a/numba/np/arraymath.py
+++ b/numba/np/arraymath.py
@@ -1886,9 +1886,9 @@ def np_partition(a, kth):
 
     kthdt = getattr(kth, 'dtype', kth)
     if numpy_version >= (2, 3):
-      kth_types = types.Integer 
+        kth_types = types.Integer
     else:
-      kth_types = (types.Boolean, types.Integer)
+        kth_types = (types.Boolean, types.Integer)
     if not isinstance(kthdt, kth_types):
         # bool gets cast to int subsequently
         raise NumbaTypeError('Partition index must be integer')
@@ -1916,9 +1916,9 @@ def np_argpartition(a, kth):
 
     kthdt = getattr(kth, 'dtype', kth)
     if numpy_version >= (2, 3):
-      kth_types = types.Integer 
+        kth_types = types.Integer
     else:
-      kth_types = (types.Boolean, types.Integer)
+        kth_types = (types.Boolean, types.Integer)
     if not isinstance(kthdt, kth_types):
         # bool gets cast to int subsequently
         raise NumbaTypeError('Partition index must be integer')

--- a/numba/tests/test_np_functions.py
+++ b/numba/tests/test_np_functions.py
@@ -2939,17 +2939,23 @@ class TestNPFunctions(MemoryLeakMixin, TestCase):
     def test_partition_boolean_inputs(self):
         pyfunc = partition
         cfunc = jit(nopython=True)(pyfunc)
+        kths = (-1, 0, 1)
+        if numpy_version < (2, 3):
+            kths = (True, False) + kths
 
         for d in np.linspace(1, 10, 17), np.array((True, False, True)):
-            for kth in True, False, -1, 0, 1:
+            for kth in kths:
                 self.partition_sanity_check(pyfunc, cfunc, d, kth)
 
     def test_argpartition_boolean_inputs(self):
         pyfunc = argpartition
         cfunc = jit(nopython=True)(pyfunc)
+        kths = (-1, 0, 1)
+        if numpy_version < (2, 3):
+            kths = (True, False) + kths
 
         for d in np.linspace(1, 10, 17), np.array((True, False, True)):
-            for kth in True, False, -1, 0, 1:
+            for kth in kths:
                 self.argpartition_sanity_check(pyfunc, cfunc, d, kth)
 
     @needs_blas

--- a/numba/tests/test_types.py
+++ b/numba/tests/test_types.py
@@ -567,7 +567,7 @@ class TestRecordDtype(unittest.TestCase):
         art1 = rec_ty[::1]
         arr = np.zeros(5, dtype=rec_dt)
         art2 = typeof(arr)
-        self.assertEqual(art2.dtype.dtype, rec_ty)
+        self.assertEqual(art2.dtype.dtype, rec_ty.dtype)
         self.assertEqual(art1, art2)
 
     def test_user_specified(self):

